design-documents/testing/functional/test-package.md

```diff
@@ -0,0 +1,104 @@
+# Magento Functional Test packaging
+The purpose of this document is to define a strategy of functional test packaging.
+
+## Terminology
+`MFTF` - Magento Functional Testing Framework
+
+`MFTF tests` - means tests written to cover Magento functionality using MFTF
+
+## Overview
+### Current situation
+In git repo all MFTF tests currently located under `app/code/Magento/*/Test/Mftf` directory.
+When Magento releasing new version we package all modules to own packages and distribute them via repo.magento.com.
+It means that all MFTF tests currently also distributed as a part of `magento2-module`.
+When SIs or Extension developers installs Magento via Composer they are able to run and use MFTF tests for their own needs.
+
+### Problem
+This approach leads to situation where production environment always contains code (in our case xml configurations) which will never be used.
+
+>There are couple reasons to separate tests from core distribution:
+>
+>Security:
+> - reduce possible attack surface (less code)
+> - the test code is not written with security in mind
+> - even if we protect the access via .htaccess, some sites run Nginx or skip .htaccess making this code accessible
+> - we had security issues in the past that resulted in remote code exec using code related to tests
+>
+>Future-proof:
+> - ability to release separately, on separate cadence
+> - matches the general plan for all other non-critical modules - we want to modularize and release separately things like payment extensions, shipping and more.
+
```

Comment:
Performance:
* MFTF coverage grows drastically, which side effect is growing deployment time for staging / production environment